src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,598 @@
+                except CallError as e:
+                    if e.errno != CallError.ENOMETHOD:
+                        raise
+
+    @private
+    def nvme_connect(self, ip, nr_io_queues=16):
+        command = ['nvme', 'connect-all', '-t', 'rdma', '-a', ip, '--persistent', '-i', f'{nr_io_queues}']
+        ret = subprocess.run(command, capture_output=True)
+        if ret.returncode:
+            error = ret.stderr.decode() if ret.stderr else ret.stdout.decode()
+            if not error:
+                error = 'No error message reported'
+            self.logger.debug('Failed to execute command: %r with error: %r', " ".join(command), error)
+            raise CallError(f'Failed connect NVMe disks: {error}')
+        return True
+
+    @private
+    def nvme_disconnect(self, ips):
+        command = ['nvme', 'list-subsys', '-o', 'json']
+        ret = subprocess.run(command, capture_output=True)
+        if ret.returncode:
+            error = ret.stderr.decode() if ret.stderr else ret.stdout.decode()
+            if not error:
+                error = 'No error message reported'
+            self.logger.debug('Failed to execute command: %r with error: %r', " ".join(command), error)
+            raise CallError(f'Failed disconnect NVMe disks: {error}')
+        data = json.loads(ret.stdout.decode())
+        for d in data:
+            for subsys in d.get('Subsystems', []):
+                matched_all_paths = True
+                for path in subsys['Paths']:
+                    matched_address = False
+                    for ip in ips:
+                        if path['Address'].startswith(f'traddr={ip},trsvcid'):
+                            matched_address = True
+                            break
+                    if not matched_address:
+                        matched_all_paths = False
+                        break
+                if matched_all_paths:
+                    command = ['nvme', 'disconnect', '-n', subsys['NQN']]
+                    ret = subprocess.run(command, capture_output=True)
+                    if ret.returncode:
+                        self.logger.debug(f'Failed to execute "{" ".join(command)}": {ret.stderr.decode()}')
+                        raise CallError(f'Failed disconnect NVMe disk: {ret.stderr.decode()}')
+
+    @private
+    async def shelf_interface_count(self, mgmt_ip):
+        try:
+            return len(await self.middleware.call('jbof.fabric_interface_choices', mgmt_ip))
+        except Exception:
+            # Really only expect 4, but we'll over-estimate for now, as we check them anyway
+            return 6
+
+    @private
+    async def unwire_host(self, mgmt_ip, shelf_index):
+        """Unware the dataplane interfaces of the specified JBOF."""
+        possible_host_ips = []
+        possible_shelf_ips = []
+        shelf_interface_count = await self.shelf_interface_count(mgmt_ip)
+        for eth_index in range(0, shelf_interface_count):
```

Comment:
This will not iterate over all of `shelf_interface_count` since range() is half-open. Are you sure this doesn't need to be `for eth_index in range(shelf_interface_count + 1)`?

Reply:
Updated to make clearer.